gio-subclass/tests/simple_application.rs

```diff
@@ -0,0 +1,168 @@
+            let prop = &PROPERTIES[id as usize];
+
+            match *prop {
+                _ => unimplemented!(),
+            }
+        }
+    }
+
+    impl ApplicationImpl<Application> for SimpleApplication{
+
+
+        fn local_command_line(&self, application: &Application, arguments: &mut ArgumentList) -> Option<i32> {
+            // application.parent_local_command_line(arguments)
+
+            for line in arguments.iter() {
+                if line.starts_with("--local-"){
+                    println!("handling argument {:?} locally\n", line);
+                }
+            }
+            arguments.remove(0);
+
+            Some(0)
+        }
+
+
+    }
+
+    struct SimpleApplicationStatic;
+
+    impl ImplTypeStatic<Application> for SimpleApplicationStatic {
+        fn get_name(&self) -> &str {
+            "SimpleApplication"
+        }
+
+        fn new(&self, obj: &Application) -> Box<ApplicationImpl<Application>> {
+            SimpleApplication::init(obj)
+        }
+
+        fn class_init(&self, klass: &mut ApplicationClass) {
+            SimpleApplication::class_init(klass);
+        }
+    }
+}
+
+glib_wrapper! {
+    //TODO: is this correct???
+    pub struct SimpleApplication(Object<imp::SimpleApplication>):
+        [Application => InstanceStruct<Application>];
+
+
+    match fn {
+        get_type => || imp::SimpleApplication::get_type().to_glib(),
+    }
+}
+
+impl SimpleApplication {
+
+    pub fn new<'a, I: Into<Option<&'a str>>>(application_id: I, flags: gio::ApplicationFlags) -> Result<SimpleApplication, glib::BoolError> {
+        use glib::object::Downcast;
+
+        unsafe {
+            match glib::Object::new(Self::static_type(), &[("application-id", &application_id.into()),
+                                                           ("flags", &flags)]){
+                Ok(obj) => Ok(obj.downcast_unchecked()),
+                Err(_) => Err(glib::BoolError("Failed to create application"))
+            }
+        }
+    }
+
+}
+
+// TODO: This one should probably get a macro
+impl ops::Deref for SimpleApplication {
+    type Target = imp::SimpleApplication;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe {
+            let base: Application = from_glib_borrow(self.to_glib_none().0);
+            let imp = base.get_impl();
+            let imp = imp.downcast_ref::<imp::SimpleApplication>().unwrap();
+            // Cast to a raw pointer to get us an appropriate lifetime: the compiler
+            // can't know that the lifetime of base is the same as the one of self
+            &*(imp as *const imp::SimpleApplication)
+        }
+    }
+}
+
+#[test]
+fn test_arguments() {
+    let app = SimpleApplication::new("org.gtk.TestApplication", gio::ApplicationFlags::empty()).unwrap();
+
+    // run is part of the trait ApplicationExtManual, so not available here!
+    // TODO: figure out how to implement that!
+    app.run(["--local"]);
```

Comment:
Doesn't this run the main loop until something else quits it again? You probably also want to check the return value